File: Novomber 2022/Day 13/C++ Code/Fill up buckets.cpp
// Problem Link :- https://practice.geeksforgeeks.org/problems/fill-up-buckets3500/1
// Video Solution Link :- https://youtu.be/RKczlfKnlEA

//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++

class Solution{
public:
    // Time Complexity :- O(n.log(n));
    // Space Complexity :- O(1);
    
    int totalWays(int n, vector<int>capacity) {
        // code here
        sort(capacity.begin(),capacity.end());
        int mod = 1000000007;
        int ans=1;
        for(int i=0;i<n;i++) {
            ans = (int)((ans*(long long int)(capacity[i]-i))%mod);
        }
        return ans;
    }
};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n;
		cin >> n;
		vector<int>capacity(n);
		for(int i = 0; i < n; i++)
			cin >> capacity[i];
		Solution ob;
		int ans = ob.totalWays(n, capacity);
		cout << ans << "\n";
	}  
	return 0;
}
// } Driver Code Ends

File: Novomber 2022/Day 13/JAVA Code/Fill up buckets.java
// Problem Link :- https://practice.geeksforgeeks.org/problems/fill-up-buckets3500/1
// Video Solution Link :- https://youtu.be/RKczlfKnlEA

//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            int n = Integer.parseInt(br.readLine().trim());
            String s = br.readLine().trim();
            String[] S = s.split(" ");
            int[] capacity = new int[n];
            for(int i = 0; i < n; i++)
                capacity[i] = Integer.parseInt(S[i]);
            Solution ob = new Solution();
            int ans = ob.totalWays(n, capacity);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution{
    // Time Complexity :- O(n.log(n));
    // Space Complexity :- O(1);
    
    public int totalWays(int n, int[] capacity) {
        // code here
        Arrays.sort(capacity);
        int ans = 1;
        final int mod = 1000000007;
        for(int i=0;i<n;i++) {
            ans = (int)((ans*(long)(capacity[i]-i))%mod);
        }
        return ans;
    }
}
